src/modalities/loss_functions.py

```diff
@@ -41,3 +41,75 @@ def __call__(self, forward_batch: InferenceResultBatch) -> torch.Tensor:
         # Flatten the tokens
         loss = self.loss_fun(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
         return loss
+
+
+class NCELoss(torch.nn.Module):
+    def __init__(self, asymm: bool, temp: float):
+        """
+        This calculates the noise contrastive estimation loss between embeddings of two different modalities
+        Implementation slightly adapted from https://arxiv.org/pdf/1912.06430.pdf by adding symmeric and asymmetric loss
+
+        Args:
+            asymm (bool): specifies if the loss is calculated in one direction or both directions
+            temp (float): temprature value for regulating loss
+        """
+        super(NCELoss, self).__init__()
+        self.asymm = asymm
+        self.temp = temp
+
+    def forward(self, embedding1: torch.Tensor, embedding2: torch.Tensor, device: torch.device) -> torch.Tensor:
+        """
+        Args:
+            embedding1 (torch.Tensor): embeddings from modality 1 of size batch x dim
+            embedding2 (torch.Tensor): embeddings from modality 2 of size batch x dim
+            device (torch.device): torch device for calculating loss
+
+        Returns:
+            torch.Tensor: loss tensor
+        """
+        # calculating the similarity matrix of size (bxb)
+        sim_matrix = torch.matmul(embedding1, embedding2.t()) / self.temp
+        # numerator of loss: using similarity scores for all positive pairs (e.g., image and its caption)
+        numerator = sim_matrix * torch.eye(sim_matrix.shape[0], device=device)
+        numerator = numerator.sum(dim=0).view(sim_matrix.shape[0], -1)
```

Comment:
could this be simplified to `torch.diag(sim_matrix, 0)`?

Reply:
`  numerator = sim_matrix * torch.eye(sim_matrix.shape[0], device=device)`
`  numerator = numerator.sum(dim=0).view(sim_matrix.shape[0], -1)`
`  numerator = torch.logsumexp(numerator, dim=1)`
So technically speaking, i think all above three lines could be replaced by `torch.diag(sim_matrix, 0)` however, I think for two reasons maybe authors wrote the code this way:

1. to keep the implementation close to the formula: 
<img width="177" alt="formula" src="https://github.com/Modalities/modalities/assets/17709204/3da69a1f-2781-4857-9479-10d337530d2c">

2. To allow for multiple instances of positive pairs per input. (MIL-NCE)

We could keep it as is and monitor if it causes any slow-down.